velox/exec/fuzzer/CacheFuzzer.cpp

```diff
@@ -0,0 +1,358 @@
+
+#include <boost/random/uniform_int_distribution.hpp>
+
+#include <folly/executors/IOThreadPoolExecutor.h>
+#include "velox/common/caching/FileIds.h"
+#include "velox/common/caching/SsdCache.h"
+#include "velox/common/file/File.h"
+#include "velox/common/file/FileSystems.h"
+#include "velox/common/memory/Memory.h"
+#include "velox/common/memory/MmapAllocator.h"
+#include "velox/dwio/common/CachedBufferedInput.h"
+#include "velox/exec/tests/utils/TempDirectoryPath.h"
+#include "velox/vector/fuzzer/VectorFuzzer.h"
+
+DEFINE_int32(steps, 10, "Number of plans to generate and test.");
+
+DEFINE_int32(
+    duration_sec,
+    0,
+    "For how long it should run (in seconds). If zero, "
+    "it executes exactly --steps iterations and exits.");
+
+DEFINE_int32(
+    max_num_reads,
+    100,
+    "Max number of reads to be performed per thread.");
+
+DEFINE_int32(num_threads, 16, "Number of read threads.");
+
+DEFINE_int32(num_source_files, 8, "Number of data files to be created.");
+
+DEFINE_uint64(
+    min_source_file_bytes,
+    32 << 20,
+    "Minimum source file size in bytes.");
+
+DEFINE_uint64(
+    max_source_file_bytes,
+    64 << 20,
+    "Maximum source file size in bytes.");
+
+DEFINE_int32(num_files_in_group, 3, "Number of files to be grouped together.");
+
+DEFINE_int64(memory_cache_bytes, 16 << 20, "Memory cache size in bytes.");
+
+DEFINE_uint64(ssd_cache_bytes, 128 << 20, "Ssd cache size in bytes.");
+
+DEFINE_int32(num_ssd_cache_shards, 4, "Number of SSD cache shards.");
+
+DEFINE_uint64(
+    ssd_checkpoint_interval_bytes,
+    64 << 20,
+    "Checkpoint after every 'ssd_checkpoint_interval_bytes'/'num_ssd_cache_shards', "
+    "written into each file. 0 means no checkpointing.");
+
+DEFINE_bool(enable_checksum, true, "Enable checksum write to SSD.");
+
+DEFINE_bool(
+    enable_checksum_read_verification,
+    true,
+    "Enable checksum read verification from SSD.");
+
+using namespace facebook::velox::cache;
+using namespace facebook::velox::dwio::common;
+
+namespace facebook::velox::exec::test {
+namespace {
+
+class CacheFuzzer {
+ public:
+  explicit CacheFuzzer(size_t seed);
+
+  void go();
+
+ private:
+  void seed(size_t seed) {
+    currentSeed_ = seed;
+    rng_.seed(currentSeed_);
+  }
+
+  void reSeed() {
+    seed(rng_());
+  }
+
+  void initSourceDataFiles();
+
+  void initializeCache();
+
+  void initializeInputs();
+
+  void readCache();
+
+  void reset();
+
+  uint64_t read(uint32_t fileIdx, int32_t fragmentIdx);
+
+  FuzzerGenerator rng_;
+  size_t currentSeed_{0};
+  std::shared_ptr<memory::MemoryPool> rootPool_{
+      memory::memoryManager()->addRootPool()};
+  std::shared_ptr<memory::MemoryPool> pool_{rootPool_->addLeafChild("leaf")};
+  std::shared_ptr<filesystems::FileSystem> fs_;
+  std::vector<std::string> fileNames_;
+  std::vector<StringIdLease> fileIds_;
+  std::vector<size_t> fileSizes_;
+  std::vector<std::vector<std::pair<int32_t, int32_t>>> fileFragments_;
```

Comment:
```
// The file fragments used to perform random reads by different threads.
//
// NOTE: the production file reader reads from the specific offset from a file instead of random location for cache reuse.
```